nomad-autoscaler sits idle without scaling jobs or evaluating policies

Hi folks,

A few times now the nomad-autoscaler (`0.3.3`) simply stops performing any actions, but the job stays up and running. There are no real logs or errors that seem to indicate the reason.

I've included my HCL and logs https://gist.github.com/evandam/c439d08a24b3d150a76335510d08a91f - note the last logs are from `2021-07-14T19:27:06.302` and the current time is `Jul 19 16:29:02 UTC 2021`. The job is still up and running, though.

You can see that up until the last log there was some activity and then it drops to zero.

![Screen Shot 2021-07-19 at 9 21 11 AM](https://user-images.githubusercontent.com/3985383/126194616-423d4a88-47a3-4355-b27b-de8714ab405e.png)

I'll drop logs to debug level to see if there's any insight, but this only seems to happen every few days/weeks sporadically.

Could it be related:

<details><summary>click to expand logs</summary>

```
2021-07-30T16:15:41.772Z [DEBUG] policy_eval.broker: ack eval: eval_id=146000dd-7f0e-c248-92ad-5507d46fccf0 token=6beec1b3-1177-9c82-f79b-54ea27f7145c eval_id=146000dd-7f0e-c248-92ad-5507d46fccf0 token=6beec1b3-1177-9c82-f79b-54ea27f7145c
2021-07-30T16:15:41.772Z [DEBUG] policy_eval.broker: eval ack'd: policy_id=88cf2aab-2148-b9b5-3291-0354924ff3da
2021-07-30T16:15:41.772Z [DEBUG] policy_eval.broker: dequeue eval: queue=horizontal
2021-07-30T16:15:41.772Z [DEBUG] policy_eval.broker: waiting for eval: queue=horizontal
2021-07-30T16:15:54.674Z [ERROR] policy_manager.policy_handler: failed to get policy: Unexpected response code: 404 (policy not found): policy_id=8dba70f3-bdfa-bb72-29a6-80c9e4e734f4
2021-07-30T16:15:54.674Z [ERROR] policy_manager.policy_handler: failed to get policy: Unexpected response code: 404 (policy not found): policy_id=88cf2aab-2148-b9b5-3291-0354924ff3da
2021-07-30T16:15:54.674Z [ERROR] policy_manager.policy_handler: failed to get policy: Unexpected response code: 404 (policy not found): policy_id=9bbdfc00-6fb1-2ecd-14c6-e41e1a5d687f
2021-07-30T16:15:54.674Z [ERROR] policy_manager.policy_handler: failed to get policy: Unexpected response code: 404 (policy not found): policy_id=cd004ff2-c6e2-e7ab-84d8-ef885cca1835
2021-07-30T16:15:54.674Z [ERROR] policy_manager.policy_handler: failed to get policy: Unexpected response code: 404 (policy not found): policy_id=9cf10d4b-bd78-10b8-15d0-aa8854c45064
2021-07-30T16:15:54.674Z [ERROR] policy_manager.policy_handler: failed to get policy: Unexpected response code: 404 (policy not found): policy_id=cd5f0198-9cf3-dec1-2e8e-6cf81aa1ba1d
2021-07-30T16:15:54.674Z [ERROR] policy_manager.policy_handler: failed to get policy: Unexpected response code: 404 (policy not found): policy_id=3044f1c6-4e0c-b215-8e80-a6bf4668bee3
2021-07-30T16:15:54.675Z [ERROR] policy_manager.policy_handler: failed to get policy: Unexpected response code: 404 (policy not found): policy_id=d69d96c5-2cd8-d8a5-c970-8ffc1bbcb7e4
2021-07-30T16:15:54.675Z [ERROR] policy_manager.policy_handler: failed to get policy: Unexpected response code: 404 (policy not found): policy_id=baa059ee-8f80-4e84-6325-2f0ee3034040
2021-07-30T16:15:54.754Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/aa787ce1-204c-d74c-ec9c-ee3af2d8b97d?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=aa787ce1-204c-d74c-ec9c-ee3af2d8b97d
2021-07-30T16:15:54.754Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/7ff593b5-71d4-c4ec-aa01-5b820d6e6b0a?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=7ff593b5-71d4-c4ec-aa01-5b820d6e6b0a
2021-07-30T16:15:54.754Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/baa059ee-8f80-4e84-6325-2f0ee3034040?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=baa059ee-8f80-4e84-6325-2f0ee3034040
2021-07-30T16:15:54.754Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/cd5f0198-9cf3-dec1-2e8e-6cf81aa1ba1d?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=cd5f0198-9cf3-dec1-2e8e-6cf81aa1ba1d
2021-07-30T16:15:54.754Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/ecc2ebdc-0615-d9f5-490d-fd39453bbbf3?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=ecc2ebdc-0615-d9f5-490d-fd39453bbbf3
2021-07-30T16:15:54.754Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/88cf2aab-2148-b9b5-3291-0354924ff3da?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=88cf2aab-2148-b9b5-3291-0354924ff3da
2021-07-30T16:15:54.754Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/9bbdfc00-6fb1-2ecd-14c6-e41e1a5d687f?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=9bbdfc00-6fb1-2ecd-14c6-e41e1a5d687f
2021-07-30T16:15:54.754Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/3044f1c6-4e0c-b215-8e80-a6bf4668bee3?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=3044f1c6-4e0c-b215-8e80-a6bf4668bee3
2021-07-30T16:15:54.754Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/712d0c71-03ab-484c-25fd-59cb974fee30?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=712d0c71-03ab-484c-25fd-59cb974fee30
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/0b03eba3-153e-4c7f-dab9-325d3cff760e?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=0b03eba3-153e-4c7f-dab9-325d3cff760e
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/d69d96c5-2cd8-d8a5-c970-8ffc1bbcb7e4?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=d69d96c5-2cd8-d8a5-c970-8ffc1bbcb7e4
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/5dc94c81-853a-a9fe-dc75-6623c696a23b?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=5dc94c81-853a-a9fe-dc75-6623c696a23b
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/cd004ff2-c6e2-e7ab-84d8-ef885cca1835?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=cd004ff2-c6e2-e7ab-84d8-ef885cca1835
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/6f01c83b-32cb-cd53-ea9d-9135e95589b7?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=6f01c83b-32cb-cd53-ea9d-9135e95589b7
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/6d513abf-f9e0-204e-c331-5b492f0ec678?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=6d513abf-f9e0-204e-c331-5b492f0ec678
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/ecb1c8b0-f8a3-1e2f-d01a-fc9041376d6a?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=ecb1c8b0-f8a3-1e2f-d01a-fc9041376d6a
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/57ba0661-2457-2b9f-af2e-cb38dd67459c?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=57ba0661-2457-2b9f-af2e-cb38dd67459c
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/8dba70f3-bdfa-bb72-29a6-80c9e4e734f4?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=8dba70f3-bdfa-bb72-29a6-80c9e4e734f4
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/cc55536e-1e39-7e6d-4728-cf4fa3cbbd49?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=cc55536e-1e39-7e6d-4728-cf4fa3cbbd49
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/6e982f0d-621b-4c11-e8e3-6794174aaf01?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=6e982f0d-621b-4c11-e8e3-6794174aaf01
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/9cf10d4b-bd78-10b8-15d0-aa8854c45064?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=9cf10d4b-bd78-10b8-15d0-aa8854c45064
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/c702348f-440f-78e7-8b92-c80161967f2a?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=c702348f-440f-78e7-8b92-c80161967f2a
2021-07-30T16:15:54.755Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/367587e3-d4c8-e9ef-2e1d-a8acba1d40ed?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=367587e3-d4c8-e9ef-2e1d-a8acba1d40ed
2021-07-30T16:15:54.756Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/d4c34548-7b25-d44d-81aa-90c40456bbf4?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=d4c34548-7b25-d44d-81aa-90c40456bbf4
2021-07-30T16:15:54.756Z [ERROR] policy_manager.policy_handler: failed to get policy: Get "http://10.82.44.27:4646/v1/scaling/policy/2c30993d-5f59-8296-7f95-23c47f28cb2c?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF: policy_id=2c30993d-5f59-8296-7f95-23c47f28cb2c
2021-07-30T16:15:54.756Z [ERROR] policy_manager: failed to call the Nomad list policies API: Get "http://10.82.44.27:4646/v1/scaling/policies?index=5967054&namespace=shared®ion=us-west-2&wait=300000ms": EOF
2021-07-30T16:15:54.756Z [DEBUG] policy_manager: re-starting policy manager
2021-07-30T16:15:56.713Z [DEBUG] internal_plugin.nomad-target: triggering run of handler garbage collection
2021-07-30T16:16:04.756Z [DEBUG] nomad_policy_source: starting policy blocking query watcher
2021-07-30T16:16:19.770Z [DEBUG] internal_plugin.nomad-target: starting job status handler: Job=automation-platform-core-be-d7b6fdf
2021-07-30T16:16:56.713Z [DEBUG] internal_plugin.nomad-target: triggering run of handler garbage collection
```
</details>

--- a/plugins/builtin/target/nomad/plugin/plugin.go
+++ b/plugins/builtin/target/nomad/plugin/plugin.go
@@ -48,11 +48,6 @@ var (
 		Name:       pluginName,
 		PluginType: sdk.PluginTypeTarget,
 	}
-
-	// statusHandlerInitTimeout is the time limit a status handler must
-	// initialize before considering the operation a failure.
-	// Declare it as a var instead of a const so we can overwrite it in tests.
-	statusHandlerInitTimeout = 30 * time.Second
 )
 
 // Assert that TargetPlugin meets the target.Target interface.
@@ -177,29 +172,14 @@ func (t *TargetPlugin) Status(config map[string]string) (*sdk.TargetStatus, erro
 
 	// Create a handler for the job if one does not currently exist.
 	if _, ok := t.statusHandlers[nsID]; !ok {
-		t.statusHandlers[nsID] = newJobScaleStatusHandler(t.client, namespace, jobID, t.logger)
-	}
-
-	// If the handler is not in a running state, start it and wait for the
-	// first run to finish.
-	if !t.statusHandlers[nsID].isRunning {
-
-		// doneCh is used to stop the goroutine in case it fails to initialize
-		// before the timeout.
-		doneCh := make(chan struct{})
-		go t.statusHandlers[nsID].start(doneCh)
-
-		// Set a timeout to make sure we don't block statusHandlers.Lock.
-		select {
-		case <-t.statusHandlers[nsID].initialDone:
-		case <-time.After(statusHandlerInitTimeout):
-			close(doneCh)
-			delete(t.statusHandlers, nsID)
-			return nil, fmt.Errorf("timeout while waiting for job scale status")
+		jsh, err := newJobScaleStatusHandler(t.client, namespace, jobID, t.logger)
+		if err != nil {
+			return nil, err
 		}
+
+		t.statusHandlers[nsID] = jsh
 	}
 
-	// Return the status data from the handler to the caller.
 	return t.statusHandlers[nsID].status(group)
 }
 
@@ -231,15 +211,7 @@ func (t *TargetPlugin) garbageCollect() {
 	defer t.statusHandlersLock.Unlock()
 
 	for jobID, handle := range t.statusHandlers {
-
-		// If the handler is running, there is no need to GC.
-		if handle.isRunning {
-			continue
-		}
-
-		// If the last updated time is before our threshold, the handler should
-		// be removed. Goodbye old friend.
-		if handle.lastUpdated < threshold {
+		if handle.shouldGC(threshold) {
 			delete(t.statusHandlers, jobID)
 			t.logger.Debug("removed inactive job status handler", "job_id", jobID)
 		}

--- a/plugins/builtin/target/nomad/plugin/plugin.go
+++ b/plugins/builtin/target/nomad/plugin/plugin.go
@@ -51,7 +51,7 @@ var (
 
 	// statusHandlerInitTimeout is the time limit a status handler must
 	// initialize before considering the operation a failure.
-	// Declare it as a var instead of a const so we can overwrite in tests.
+	// Declare it as a var instead of a const so we can overwrite it in tests.
 	statusHandlerInitTimeout = 30 * time.Second
 )
 

--- a/plugins/builtin/target/nomad/plugin/plugin.go
+++ b/plugins/builtin/target/nomad/plugin/plugin.go
@@ -211,6 +211,7 @@ func (t *TargetPlugin) garbageCollect() {
 
 	// Iterate all the handlers, ensuring we lock for safety.
 	t.statusHandlersLock.Lock()
+	defer t.statusHandlersLock.Unlock()
 
 	for jobID, handle := range t.statusHandlers {
 
@@ -226,6 +227,4 @@ func (t *TargetPlugin) garbageCollect() {
 			t.logger.Debug("removed inactive job status handler", "job_id", jobID)
 		}
 	}
-
-	t.statusHandlersLock.Unlock()
 }

--- a/plugins/builtin/target/nomad/plugin/plugin.go
+++ b/plugins/builtin/target/nomad/plugin/plugin.go
@@ -48,6 +48,11 @@ var (
 		Name:       pluginName,
 		PluginType: sdk.PluginTypeTarget,
 	}
+
+	// statusHandlerInitTimeout is the time limit a status handler must
+	// initialize before considering the operation a failure.
+	// Declare it as a var instead of a const so we can overwrite in tests.
+	statusHandlerInitTimeout = 30 * time.Second
 )
 
 // Assert that TargetPlugin meets the target.Target interface.
@@ -178,8 +183,20 @@ func (t *TargetPlugin) Status(config map[string]string) (*sdk.TargetStatus, erro
 	// If the handler is not in a running state, start it and wait for the
 	// first run to finish.
 	if !t.statusHandlers[nsID].isRunning {
-		go t.statusHandlers[nsID].start()
-		<-t.statusHandlers[nsID].initialDone
+
+		// doneCh is used to stop the goroutine in case it fails to initialize
+		// before the timeout.
+		doneCh := make(chan struct{})
+		go t.statusHandlers[nsID].start(doneCh)
+
+		// Set a timeout to make sure we don't block statusHandlers.Lock.
+		select {
+		case <-t.statusHandlers[nsID].initialDone:
+		case <-time.After(statusHandlerInitTimeout):
+			close(doneCh)
+			delete(t.statusHandlers, nsID)
+			return nil, fmt.Errorf("timeout while waiting for job scale status")
+		}
 	}
 
 	// Return the status data from the handler to the caller.

--- a/plugins/builtin/target/nomad/plugin/plugin_test.go
+++ b/plugins/builtin/target/nomad/plugin/plugin_test.go
@@ -3,12 +3,14 @@ package nomad
 import (
 	"net/http"
 	"net/http/httptest"
+	"sync"
 	"testing"
 	"time"
 
 	hclog "github.com/hashicorp/go-hclog"
 	"github.com/hashicorp/nomad-autoscaler/sdk"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 func TestTargetPlugin_garbageCollect(t *testing.T) {
@@ -43,8 +45,49 @@ func TestTargetPlugin_garbageCollect(t *testing.T) {
 	})
 }
 
-func TestTargetPlugin_statusTimeout(t *testing.T) {
+func TestTargetPlugin_Status(t *testing.T) {
 	nomadMock := httptest.NewServer(http.HandlerFunc(scaleStatusHandler))
+	defer nomadMock.Close()
+
+	plugin := PluginConfig.Factory(hclog.NewNullLogger()).(*TargetPlugin)
+	plugin.SetConfig(map[string]string{
+		"nomad_address": nomadMock.URL,
+	})
+
+	expected := &sdk.TargetStatus{
+		Ready: true,
+		Count: 0,
+		Meta: map[string]string{
+			"nomad_autoscaler.target.nomad.example.stopped": "false",
+		},
+	}
+	got, err := plugin.Status(map[string]string{
+		"Job":       "example",
+		"Group":     "cache",
+		"Namespace": "default",
+	})
+	require.NoError(t, err)
+	assert.Equal(t, expected, got)
+
+	// Call Status multiple times concurrently to test for data races.
+	var wg sync.WaitGroup
+	for i := 0; i < 100; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			_, err := plugin.Status(map[string]string{
+				"Job":       "example",
+				"Group":     "cache",
+				"Namespace": "default",
+			})
+			assert.NoError(t, err)
+		}()
+	}
+	wg.Wait()
+}
+
+func TestTargetPlugin_statusTimeout(t *testing.T) {
+	nomadMock := httptest.NewServer(http.HandlerFunc(scaleStatusErrorHandler))
 	defer nomadMock.Close()
 
 	statusHandlerInitTimeout = 3 * time.Second
@@ -77,5 +120,27 @@ func TestTargetPlugin_statusTimeout(t *testing.T) {
 }
 
 func scaleStatusHandler(w http.ResponseWriter, r *http.Request) {
+	respBody := `
+{
+  "JobCreateIndex": 10,
+  "JobID": "example",
+  "Namespace": "default",
+  "JobModifyIndex": 18,
+  "JobStopped": false,
+  "TaskGroups": {
+    "cache": {
+      "Desired": 1,
+      "Events": null,
+      "Healthy": 1,
+      "Placed": 1,
+      "Running": 0,
+      "Unhealthy": 0
+    }
+  }
+}`
+	w.Write([]byte(respBody))
+}
+
+func scaleStatusErrorHandler(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusInternalServerError)
 }

--- a/plugins/builtin/target/nomad/plugin/plugin_test.go
+++ b/plugins/builtin/target/nomad/plugin/plugin_test.go
@@ -1,10 +1,13 @@
 package nomad
 
 import (
+	"net/http"
+	"net/http/httptest"
 	"testing"
 	"time"
 
 	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad-autoscaler/sdk"
 	"github.com/stretchr/testify/assert"
 )
 
@@ -17,12 +20,12 @@ func TestTargetPlugin_garbageCollect(t *testing.T) {
 	targetPlugin := TargetPlugin{
 		logger: hclog.NewNullLogger(),
 		statusHandlers: map[namespacedJobID]*jobScaleStatusHandler{
-			namespacedJobID{"default", "running"}:               {isRunning: true, lastUpdated: curTime},
-			namespacedJobID{"default", "recently-stopped"}:      {isRunning: false, lastUpdated: curTime - 1800000000000},
-			namespacedJobID{"default", "stopped-long-time-ago"}: {isRunning: false, lastUpdated: curTime - 18000000000000},
-			namespacedJobID{"special", "running"}:               {isRunning: true, lastUpdated: curTime},
-			namespacedJobID{"special", "recently-stopped"}:      {isRunning: false, lastUpdated: curTime - 1800000000000},
-			namespacedJobID{"special", "stopped-long-time-ago"}: {isRunning: false, lastUpdated: curTime - 18000000000000},
+			{"default", "running"}:               {isRunning: true, lastUpdated: curTime},
+			{"default", "recently-stopped"}:      {isRunning: false, lastUpdated: curTime - 1800000000000},
+			{"default", "stopped-long-time-ago"}: {isRunning: false, lastUpdated: curTime - 18000000000000},
+			{"special", "running"}:               {isRunning: true, lastUpdated: curTime},
+			{"special", "recently-stopped"}:      {isRunning: false, lastUpdated: curTime - 1800000000000},
+			{"special", "stopped-long-time-ago"}: {isRunning: false, lastUpdated: curTime - 18000000000000},
 		},
 	}
 
@@ -38,4 +41,41 @@ func TestTargetPlugin_garbageCollect(t *testing.T) {
 		assert.NotNil(t, targetPlugin.statusHandlers[namespacedJobID{"special", "recently-stopped"}], testName)
 		assert.Len(t, targetPlugin.statusHandlers, 4, testName)
 	})
+}
+
+func TestTargetPlugin_statusTimeout(t *testing.T) {
+	nomadMock := httptest.NewServer(http.HandlerFunc(scaleStatusHandler))
+	defer nomadMock.Close()
+
+	statusHandlerInitTimeout = 3 * time.Second
+
+	plugin := PluginConfig.Factory(hclog.NewNullLogger()).(*TargetPlugin)
+	plugin.SetConfig(map[string]string{
+		"nomad_address": nomadMock.URL,
+	})
+
+	var statusErr error
+	var status *sdk.TargetStatus
+	doneCh := make(chan struct{})
+	go func() {
+		status, statusErr = plugin.Status(map[string]string{
+			"Job":       "example",
+			"Group":     "cache",
+			"Namespace": "default",
+		})
+		close(doneCh)
+	}()
+
+	select {
+	case <-doneCh:
+	case <-time.After(2 * statusHandlerInitTimeout):
+		t.Fatalf("status call blocked")
+	}
+
+	assert.Error(t, statusErr)
+	assert.Nil(t, status)
+}
+
+func scaleStatusHandler(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusInternalServerError)
 }

--- a/plugins/builtin/target/nomad/plugin/state.go
+++ b/plugins/builtin/target/nomad/plugin/state.go
@@ -23,6 +23,13 @@ const (
 	metaKeyJobStoppedSuffix = ".stopped"
 )
 
+var (
+	// statusHandlerInitTimeout is the time limit a status handler must
+	// initialize before considering the operation a failure.
+	// Declared as a var instead of a const to allow overwriting it in tests.
+	statusHandlerInitTimeout = 30 * time.Second
+)
+
 // jobScaleStatusHandler is an individual handler on the /v1/job/<job>/scale
 // GET endpoint. It provides methods for obtaining the current scaling state of
 // a job and task group.
@@ -55,14 +62,27 @@ type jobScaleStatusHandler struct {
 	lastUpdated int64
 }
 
-func newJobScaleStatusHandler(client *api.Client, ns, jobID string, logger hclog.Logger) *jobScaleStatusHandler {
-	return &jobScaleStatusHandler{
+func newJobScaleStatusHandler(client *api.Client, ns, jobID string, logger hclog.Logger) (*jobScaleStatusHandler, error) {
+	jsh := &jobScaleStatusHandler{
 		client:      client,
 		initialDone: make(chan bool),
 		jobID:       jobID,
 		namespace:   ns,
 		logger:      logger.With(configKeyJobID, jobID),
 	}
+
+	go jsh.start()
+
+	// Wait for initial status data to be loaded.
+	// Set a timeout to make sure callers are not blocked indefinetely.
+	select {
+	case <-jsh.initialDone:
+	case <-time.After(statusHandlerInitTimeout):
+		jsh.setStopState()
+		return nil, fmt.Errorf("timeout while waiting for job scale status handler")
+	}
+
+	return jsh, nil
 }
 
 // status returns the cached scaling status of the passed group.
@@ -126,7 +146,7 @@ func (jsh *jobScaleStatusHandler) status(group string) (*sdk.TargetStatus, error
 
 // start runs the blocking query loop that processes changes from the API and
 // reflects the status internally.
-func (jsh *jobScaleStatusHandler) start(doneCh <-chan struct{}) {
+func (jsh *jobScaleStatusHandler) start() {
 
 	// Log that we are starting, useful for debugging.
 	jsh.logger.Debug("starting job status handler")
@@ -142,14 +162,11 @@ func (jsh *jobScaleStatusHandler) start(doneCh <-chan struct{}) {
 	}
 
 	for {
-		select {
-		case <-doneCh:
-			jsh.setStopState()
-			return
-		default:
-		}
-
 		status, meta, err := jsh.client.Jobs().ScaleStatus(jsh.jobID, q)
+
+		// Update the handlers state.
+		jsh.updateStatusState(status, err)
+
 		if err != nil {
 			jsh.logger.Debug("failed to read job scale status, retrying", "error", err)
 
@@ -160,7 +177,6 @@ func (jsh *jobScaleStatusHandler) start(doneCh <-chan struct{}) {
 				jsh.setStopState()
 				return
 			}
-			jsh.updateStatusState(status, err)
 
 			// Reset query WaitIndex to zero so we can get the job status
 			// immediately in the next request instead of blocking and having
@@ -173,16 +189,24 @@ func (jsh *jobScaleStatusHandler) start(doneCh <-chan struct{}) {
 			continue
 		}
 
+		// Read handler state into local variables so we don't starve the lock.
+		jsh.lock.RLock()
+		isRunning := jsh.isRunning
+		scaleStatus := jsh.scaleStatus
+		jsh.lock.RUnlock()
+
+		// Stop loop if handler is not running anymore.
+		if !isRunning {
+			return
+		}
+
 		// If the index has not changed, the query returned because the timeout
 		// was reached, therefore start the next query loop.
 		// The index could also be the same when a reconnect happens, in which
 		// case the handler state needs to be updated regardless of the index.
-		if jsh.scaleStatus != nil && !blocking.IndexHasChanged(meta.LastIndex, q.WaitIndex) {
+		if scaleStatus != nil && !blocking.IndexHasChanged(meta.LastIndex, q.WaitIndex) {
 			continue
 		}
-
-		// Update the handlers state.
-		jsh.updateStatusState(status, nil)
 
 		// Modify the wait index on the QueryOptions so the blocking query
 		// is using the latest index value.
@@ -220,4 +244,12 @@ func (jsh *jobScaleStatusHandler) setStopState() {
 	jsh.scaleStatus = nil
 	jsh.scaleStatusError = nil
 	jsh.lastUpdated = time.Now().UTC().UnixNano()
+}
+
+// shouldGC returns true if the handler is not considered as active anymore.
+func (jsh *jobScaleStatusHandler) shouldGC(threshold int64) bool {
+	jsh.lock.RLock()
+	defer jsh.lock.RUnlock()
+
+	return !jsh.isRunning && jsh.lastUpdated < threshold
 }

--- a/plugins/builtin/target/nomad/plugin/state.go
+++ b/plugins/builtin/target/nomad/plugin/state.go
@@ -4,6 +4,7 @@ import (
 	"fmt"
 	"strconv"
 	"strings"
+	"sync"
 	"time"
 
 	hclog "github.com/hashicorp/go-hclog"
@@ -31,6 +32,9 @@ type jobScaleStatusHandler struct {
 
 	namespace string
 	jobID     string
+
+	// lock is used to synchronize access to the status variables below.
+	lock sync.RWMutex
 
 	// scaleStatus is the internal reflection of the response objects from the
 	// job scale status API.
@@ -63,6 +67,8 @@ func newJobScaleStatusHandler(client *api.Client, ns, jobID string, logger hclog
 
 // status returns the cached scaling status of the passed group.
 func (jsh *jobScaleStatusHandler) status(group string) (*sdk.TargetStatus, error) {
+	jsh.lock.RLock()
+	defer jsh.lock.RUnlock()
 
 	// If the last status response included an error, just return this to the
 	// caller.
@@ -124,7 +130,10 @@ func (jsh *jobScaleStatusHandler) start(doneCh <-chan struct{}) {
 
 	// Log that we are starting, useful for debugging.
 	jsh.logger.Debug("starting job status handler")
+
+	jsh.lock.Lock()
 	jsh.isRunning = true
+	jsh.lock.Unlock()
 
 	q := &api.QueryOptions{
 		Namespace: jsh.namespace,
@@ -184,6 +193,8 @@ func (jsh *jobScaleStatusHandler) start(doneCh <-chan struct{}) {
 // updateStatusState takes the API responses and updates the internal state
 // along with a timestamp.
 func (jsh *jobScaleStatusHandler) updateStatusState(status *api.JobScaleStatusResponse, err error) {
+	jsh.lock.Lock()
+	defer jsh.lock.Unlock()
 
 	// Mark the handler as initialized and notify initialDone channel.
 	if !jsh.initialized {
@@ -202,6 +213,9 @@ func (jsh *jobScaleStatusHandler) updateStatusState(status *api.JobScaleStatusRe
 // setStopState handles updating state when the job status handler is going to
 // stop.
 func (jsh *jobScaleStatusHandler) setStopState() {
+	jsh.lock.Lock()
+	defer jsh.lock.Unlock()
+
 	jsh.isRunning = false
 	jsh.scaleStatus = nil
 	jsh.scaleStatusError = nil

--- a/plugins/builtin/target/nomad/plugin/state.go
+++ b/plugins/builtin/target/nomad/plugin/state.go
@@ -74,7 +74,7 @@ func newJobScaleStatusHandler(client *api.Client, ns, jobID string, logger hclog
 	go jsh.start()
 
 	// Wait for initial status data to be loaded.
-	// Set a timeout to make sure callers are not blocked indefinetely.
+	// Set a timeout to make sure callers are not blocked indefinitely.
 	select {
 	case <-jsh.initialDone:
 	case <-time.After(statusHandlerInitTimeout):

--- a/plugins/builtin/target/nomad/plugin/state.go
+++ b/plugins/builtin/target/nomad/plugin/state.go
@@ -120,7 +120,7 @@ func (jsh *jobScaleStatusHandler) status(group string) (*sdk.TargetStatus, error
 
 // start runs the blocking query loop that processes changes from the API and
 // reflects the status internally.
-func (jsh *jobScaleStatusHandler) start() {
+func (jsh *jobScaleStatusHandler) start(doneCh <-chan struct{}) {
 
 	// Log that we are starting, useful for debugging.
 	jsh.logger.Debug("starting job status handler")
@@ -133,8 +133,16 @@ func (jsh *jobScaleStatusHandler) start() {
 	}
 
 	for {
+		select {
+		case <-doneCh:
+			jsh.setStopState()
+			return
+		default:
+		}
+
 		status, meta, err := jsh.client.Jobs().ScaleStatus(jsh.jobID, q)
 		if err != nil {
+			jsh.logger.Debug("failed to read job scale status, retrying", "error", err)
 
 			// If the job is not found on the cluster, stop the handlers loop
 			// process and set terminal state. It is still possible to read the
@@ -167,26 +175,25 @@ func (jsh *jobScaleStatusHandler) start() {
 		// Update the handlers state.
 		jsh.updateStatusState(status, nil)
 
-		// Mark the handler as initialized and notify initialDone channel.
-		if !jsh.initialized {
-			jsh.handleFirstRun()
-			jsh.initialized = true
-		}
-
 		// Modify the wait index on the QueryOptions so the blocking query
 		// is using the latest index value.
 		q.WaitIndex = meta.LastIndex
 	}
 }
 
-// handleFirstRun is a helper function which responds to channel listeners that
-// the first run of the blocking query has completed and therefore data is
-// available for querying.
-func (jsh *jobScaleStatusHandler) handleFirstRun() { jsh.initialDone <- true }
-
 // updateStatusState takes the API responses and updates the internal state
 // along with a timestamp.
 func (jsh *jobScaleStatusHandler) updateStatusState(status *api.JobScaleStatusResponse, err error) {
+
+	// Mark the handler as initialized and notify initialDone channel.
+	if !jsh.initialized {
+		jsh.initialized = true
+
+		// Close channel so we don't block waiting for readers.
+		// jsh.initialized should only be set once to avoid closing this twice.
+		close(jsh.initialDone)
+	}
+
 	jsh.scaleStatus = status
 	jsh.scaleStatusError = err
 	jsh.lastUpdated = time.Now().UTC().UnixNano()

--- a/plugins/builtin/target/nomad/plugin/state_test.go
+++ b/plugins/builtin/target/nomad/plugin/state_test.go
@@ -8,6 +8,7 @@ import (
 	"github.com/hashicorp/nomad-autoscaler/sdk"
 	"github.com/hashicorp/nomad/api"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 func Test_newJobStateHandler(t *testing.T) {
@@ -17,7 +18,9 @@ func Test_newJobStateHandler(t *testing.T) {
 	assert.Nil(t, err)
 
 	// Create the new handler and perform assertions.
-	jsh := newJobScaleStatusHandler(c, "default", "test", hclog.NewNullLogger())
+	jsh, err := newJobScaleStatusHandler(c, "default", "test", hclog.NewNullLogger())
+	require.NoError(t, err)
+
 	assert.NotNil(t, jsh.client)
 	assert.Equal(t, "test", jsh.jobID)
 	assert.NotNil(t, jsh.initialDone)
@@ -111,10 +114,8 @@ func Test_jobStateHandler_status(t *testing.T) {
 }
 
 func Test_jobStateHandler_updateStatusState(t *testing.T) {
-	c, err := api.NewClient(api.DefaultConfig())
-	assert.Nil(t, err)
-
-	jsh := newJobScaleStatusHandler(c, "", "", hclog.NewNullLogger())
+	jsh := &jobScaleStatusHandler{}
+	jsh.initialDone = make(chan bool)
 
 	// Assert that the lastUpdated timestamp is default. This helps confirm it
 	// gets updated later in the test.

--- a/plugins/builtin/target/nomad/plugin/state_test.go
+++ b/plugins/builtin/target/nomad/plugin/state_test.go
@@ -111,7 +111,10 @@ func Test_jobStateHandler_status(t *testing.T) {
 }
 
 func Test_jobStateHandler_updateStatusState(t *testing.T) {
-	jsh := &jobScaleStatusHandler{}
+	c, err := api.NewClient(api.DefaultConfig())
+	assert.Nil(t, err)
+
+	jsh := newJobScaleStatusHandler(c, "", "", hclog.NewNullLogger())
 
 	// Assert that the lastUpdated timestamp is default. This helps confirm it
 	// gets updated later in the test.

--- a/policy/handler.go
+++ b/policy/handler.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"math/rand"
 	"strconv"
 	"sync"
 	"time"
@@ -309,6 +310,12 @@ func (h *Handler) updateHandler(current, next *sdk.ScalingPolicy) {
 	// policy's evaluation interval has changed.
 	if current == nil || current.EvaluationInterval != next.EvaluationInterval {
 		h.ticker.Stop()
+
+		// Add a small random delay to spread the first evaluation of policies
+		// that are evaluated at the same time.
+		splayNs := rand.Intn(30) * 100 * 1000 * 1000
+		time.Sleep(time.Duration(splayNs))
+
 		h.ticker = time.NewTicker(next.EvaluationInterval)
 	}
 }

--- a/policy/handler.go
+++ b/policy/handler.go
@@ -311,8 +311,8 @@ func (h *Handler) updateHandler(current, next *sdk.ScalingPolicy) {
 	if current == nil || current.EvaluationInterval != next.EvaluationInterval {
 		h.ticker.Stop()
 
-		// Add a small random delay to spread the first evaluation of policies
-		// that are evaluated at the same time.
+		// Add a small random delay between 0 and 300ms to spread the first
+		// evaluation of policies that are loaded at the same time.
 		splayNs := rand.Intn(30) * 100 * 1000 * 1000
 		time.Sleep(time.Duration(splayNs))
 

